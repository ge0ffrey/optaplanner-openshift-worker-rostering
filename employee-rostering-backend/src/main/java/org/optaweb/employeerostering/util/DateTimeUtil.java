/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.optaweb.employeerostering.util;

import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.ZoneId;

public class DateTimeUtil {

    public static LocalDateTime toLocalDateTimeInZone(OffsetDateTime dateTime, ZoneId zoneId) {
        return LocalDateTime.ofEpochSecond(dateTime.toEpochSecond(), dateTime.getNano(),
                                           zoneId.getRules().getOffset(dateTime.toInstant()));
    }
}
