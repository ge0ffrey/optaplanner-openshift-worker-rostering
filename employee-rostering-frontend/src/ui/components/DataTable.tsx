/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  Table,
  TableHeader,
  TableBody,
  headerCol,
  IRow
} from '@patternfly/react-table';
import {Button, ButtonVariant} from '@patternfly/react-core';
import { SaveIcon, CloseIcon, EditIcon, TrashIcon } from '@patternfly/react-icons';
import { EditableComponent } from './EditableComponent';

export interface DataTableProps<T> {
  title: string;
  columnTitles: string[];
  tableData: T[];
}

interface DataTableState<T> {
  newRowData: Partial<T>|null;
  currentFilter: (rowData: T) => boolean;
}

export type ReadonlyPartial<T> = { readonly [P in keyof T]?: T[P] };
export type PropertySetter<T> = (propertyName: keyof T, value: T[keyof T]|undefined) => void;

export abstract class DataTable<T, P extends DataTableProps<T>> extends React.Component<P, DataTableState<T>> {
  constructor(props: P) {
    super(props);
    this.createNewRow = this.createNewRow.bind(this);
    this.cancelAddingRow = this.cancelAddingRow.bind(this);
    this.convertDataToTableRow = this.convertDataToTableRow.bind(this);
    this.state = {newRowData: null, currentFilter: (t) => true};
  }

  abstract displayDataRow(data: T): JSX.Element[];
  abstract editDataRow(data: ReadonlyPartial<T>, setProperty: (propertyName: keyof T, value: T[keyof T]|undefined) => void): JSX.Element[];
  abstract isValid(editedValue: ReadonlyPartial<T>): editedValue is T;

  abstract updateData(data: T): void;
  abstract addData(data: T): void;
  abstract removeData(data: T): void;

  createNewRow() {
    if (this.state.newRowData === null) {
      this.setState({...this.state, newRowData: {}});
    }
  }

  cancelAddingRow() {
    this.setState({...this.state, newRowData: null});
  }

  getAddButtons(newData: Partial<T>): JSX.Element {
    return <span>
      <Button aria-label="Save"
        variant={ButtonVariant.link}
        onClick={() => {
          if (this.isValid(newData)) {
            this.addData(newData);
            this.cancelAddingRow();
          }
        }}>
        <SaveIcon />
      </Button>
      <Button aria-label="Cancel"
        variant={ButtonVariant.link}
        onClick={this.cancelAddingRow}>
        <CloseIcon />
      </Button>
    </span>;
  }

  getEditButtons(originalData: T, editedData: Partial<T>, editableComponents: EditableComponent[]): JSX.Element {
    return <EditableComponent ref={(c) => {editableComponents[editableComponents.length-1] = c as EditableComponent;}}
      viewer={<span>
        <Button aria-label="Edit"
          variant={ButtonVariant.link}
          onClick={() => {
            editableComponents.forEach(c => c.startEditing());
          }}>
          <EditIcon />
        </Button>
        <Button aria-label="Delete"
          variant={ButtonVariant.link}
          onClick={() => {
            this.removeData(originalData);
          }}>
          <TrashIcon />
        </Button>
      </span>}
      editor={<span>
        <Button aria-label="Save"
          variant={ButtonVariant.link}
          onClick={() => {
            if (this.isValid(editedData)) {
              this.updateData(editedData);
              editableComponents.forEach(c => c.stopEditing());
            }
          }}>
          <SaveIcon />
        </Button>
        <Button aria-label="Cancel"
          variant={ButtonVariant.link}
          onClick={() => {
            editableComponents.forEach(c => c.stopEditing());
          }}>
          <CloseIcon />
        </Button>
      </span>}/>;
  }

  convertDataToTableRow(data: T): IRow {
    const editedData: Partial<T> = {...data};
    const viewers = this.displayDataRow(data);

    const setProperty = (key: keyof T, value: T[keyof T]|undefined) => editedData[key] = value;
    const editors = this.editDataRow(editedData, setProperty);
    const length = viewers.length
    const editableComponents: EditableComponent[] = new Array(length + 1);
    const cellContents = viewers.map((viewer, index) => {return { title:
      <EditableComponent viewer={viewer}
        editor={editors[index]}
        ref={(c) => editableComponents[index] = c as EditableComponent}
      />} }).concat([{title: this.getEditButtons(data, editedData, editableComponents)}]);
    return {
      cells: cellContents
    };
  }

  // Use for SNAPSHOT testing
  renderViewer(data: T): JSX.Element {
    return <tr>{this.displayDataRow(data).map((c,index) => <td key={index}>{c}</td>)}</tr>;
  }

  // Use for SNAPSHOT testing
  renderEditor(data: T): JSX.Element {
    const editedData: T = {...data};
    const setProperty = (key: keyof T, value: any) => editedData[key] = value;
    return <tr>{this.editDataRow(editedData, setProperty).map((c,index) => <td key={index}>{c}</td>)}</tr>;
  }

  render() {
    const setProperty = (key: keyof T,value: T[keyof T]|undefined) => this.setState(prevState => 
      ({...prevState, newRowData: {...prevState.newRowData, [key]: value}}));
    const additionalRows: IRow[] = (this.state.newRowData !== null)?
      [
        {
          cells: this.editDataRow(this.state.newRowData, setProperty).map(c => {return {title: c}})
            .concat([{title: this.getAddButtons(this.state.newRowData)}])
        }
      ] : [];
    const rows = additionalRows.concat(this.props.tableData
      .filter(this.state.currentFilter).map(this.convertDataToTableRow));
    const columns = this.props.columnTitles.map(t => { 
      return { title: t, cellTransforms: [headerCol], props: {} };
    }).concat([{title: '', cellTransforms: [headerCol], props: {}}]);
    return (
      <div>
        <Button isDisabled={this.state.newRowData !== null} onClick={this.createNewRow}>Add</Button>
        <Table caption={this.props.title} cells={columns} rows={rows}>
          <TableHeader />
          <TableBody />
        </Table>
      </div>
    );
  }
}

export default DataTable
