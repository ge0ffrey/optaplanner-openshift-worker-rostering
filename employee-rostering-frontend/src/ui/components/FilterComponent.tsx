/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from "react";
import { TextInput, Button, ButtonVariant } from "@patternfly/react-core";
import './FilterComponent.css';
import { Predicate } from "types";

export interface FilterProps<T> {
  filter: (filter: string) => Predicate<T>;
  onChange: (filter: Predicate<T>) => void;
}

export interface FilterState {
  filterText: string;
}

class FilterComponent<T> extends React.Component<FilterProps<T>, FilterState> {

  constructor(props: FilterProps<T>) {
    super(props);
    this.state = {filterText: ""};
    this.updateFilter = this.updateFilter.bind(this);
  }

  updateFilter(filterText: string) {
    this.props.onChange(this.props.filter(filterText));
    this.setState({filterText});
  }

  render() {
    return (
      <div className="search-icons">
        <TextInput
          aria-label="Search"
          placeholder="Search..."
          value={this.state.filterText}
          onChange={this.updateFilter} 
        />
        <Button
          variant={ButtonVariant.plain}
          isDisabled={this.state.filterText.length === 0} 
          onClick={() => this.updateFilter("")}
        >
          <svg style={{verticalAlign: "-0.125em"}} fill="currentColor" height="1em" width="1em" viewBox="0 0 512 512" aria-hidden="true" role="img">
            <path d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z" transform="" />
          </svg>
        </Button>
      </div>
    );
  }
}

export default FilterComponent;