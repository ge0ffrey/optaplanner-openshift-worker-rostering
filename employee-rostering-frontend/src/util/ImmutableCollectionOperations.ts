/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import DomainObject from 'domain/DomainObject';

export function toggleElement<T>(collection: T[], element: T): T[] {
  if (collection.indexOf(element) !== -1) {
    return collection.filter(e => e !== element);
  }
  else {
    return [...collection, element];
  }
}

export function withoutElement<T extends DomainObject>(collection: T[], removedElement: T): T[] {
  return collection.filter(element => element.id !== removedElement.id);
}

export function withElement<T extends DomainObject>(collection: T[], addedElement: T): T[] {
  return [...collection, addedElement];
}

export function withUpdatedElement<T extends DomainObject>(collection: T[], updatedElement: T): T[] {
  return withElement(withoutElement(collection, updatedElement), updatedElement);
}

export function mapWithoutElement<T extends DomainObject>(map: Map<number, T>, removedElement: T): Map<number, T> {
  return collection.filter(element => element.id !== removedElement.id);
}

export function mapWithElement<T extends DomainObject>(map: Map<number, T>, addedElement: T): Map<number, T> {
  return [...collection, addedElement];
}

export function mapWithUpdatedElement<T extends DomainObject>(map: Map<number, T>, updatedElement: T): Map<number, T> {
  return withElement(withoutElement(collection, updatedElement), updatedElement);
}

