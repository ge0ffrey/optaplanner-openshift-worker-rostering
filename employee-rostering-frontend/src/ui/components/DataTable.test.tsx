/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { DataTable, DataTableProps } from './DataTable';
import { stringSorter} from 'util/CommonSorters';
import FilterComponent from './FilterComponent';
import { useTranslation } from 'react-i18next';

interface MockData {name: string; number: number}
class MockDataTable extends DataTable<MockData, DataTableProps<MockData>> {
  displayDataRow = jest.fn((data) => [<span key={0} id="viewer">{data.name}</span>,
    <span key={1}>{data.number}</span>]);
  getInitialStateForNewRow = jest.fn(() => ({}));
  editDataRow = jest.fn((data) => [<input key={0} id="editor" />,
    <input key={1} />]);
  isValid = jest.fn();
  isDataComplete = jest.fn() as any;
  getSorters = jest.fn(() => [null, stringSorter((d: MockData) => String(d.number))]);
  getFilter = jest.fn((() => (filter: string) => (t: MockData) => Boolean(true)));

  updateData = jest.fn();
  addData = jest.fn();
  removeData = jest.fn();
}

describe('DataTable component', () => {
  it('should render correctly with no rows', () => {
    const dataTable = shallow(<MockDataTable {...noRows} />);
    expect(toJson(dataTable)).toMatchSnapshot();
  });

  it('should render correctly with a few rows', () => {
    const dataTable = shallow(<MockDataTable {...twoRows} />);
    expect(toJson(dataTable)).toMatchSnapshot();
  });

  it('should call display data row and edit data row for each row in render', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.render();
    expect(dataTable.displayDataRow).toBeCalledTimes(2);
    expect(dataTable.displayDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[0]);
    expect(dataTable.displayDataRow).toHaveBeenNthCalledWith(2, twoRows.tableData[1]);
    expect(dataTable.editDataRow).toBeCalledTimes(2);
    expect(dataTable.editDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[0], expect.any(Function));
    expect(dataTable.editDataRow).toHaveBeenNthCalledWith(2, twoRows.tableData[1], expect.any(Function));
  });

  it('should render viewer initially', () => {
    const dataTable = new MockDataTable(twoRows);
    const table = shallow(<div>{dataTable.render()}</div>);
    expect(toJson(table)).toMatchSnapshot();
  });

  it('should set new row data to intial state if no row is being added', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.getInitialStateForNewRow.mockReturnValue({ name: "Hi" });
    dataTable.setState = jest.fn();
    dataTable.createNewRow();

    expect(dataTable.getInitialStateForNewRow).toBeCalled();
    expect(dataTable.setState).toBeCalled();
    expect(dataTable.setState).toBeCalledWith({ newRowData: {name: "Hi" } });
  });

  it('should not set new row data to intial state if no row is being added', () => {
    const dataTable = mount(<MockDataTable {...twoRows} />);
    dataTable.setState({ newRowData: { name: "Hi" } });
    (dataTable.instance() as MockDataTable).setState = jest.fn();
    (dataTable.instance() as MockDataTable).createNewRow();

    expect((dataTable.instance() as MockDataTable).getInitialStateForNewRow).not.toBeCalled();
    expect((dataTable.instance() as MockDataTable).setState).not.toBeCalled();
  });

  it('should render editor for new row', () => {
    const dataTable = mount(<MockDataTable {...twoRows} />);
    (dataTable.instance() as MockDataTable).displayDataRow.mockClear();
    (dataTable.instance() as MockDataTable).editDataRow.mockClear();
    (dataTable.instance() as MockDataTable).createNewRow();

    expect((dataTable.instance() as MockDataTable).getInitialStateForNewRow).toBeCalled();
    expect((dataTable.instance() as MockDataTable).displayDataRow).toBeCalledTimes(2);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[0]);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(2, twoRows.tableData[1]);
    expect((dataTable.instance() as MockDataTable).editDataRow).toBeCalledTimes(3);
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(1, {}, expect.any(Function));
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(2, twoRows.tableData[0], expect.any(Function));
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(3, twoRows.tableData[1], expect.any(Function));
    expect(toJson(shallow(<div>{dataTable.instance().render()}</div>))).toMatchSnapshot();
  });

  it('clicking edit button should edit row', () => {
    const dataTable = new MockDataTable(twoRows);
    const data: MockData = {name: "Hello", number: 1};
    const toggleEditing = jest.fn();
    const buttons = mount(dataTable.getEditButtons(data, {...data}, false, toggleEditing));
    buttons.find('button[aria-label="Edit"]').simulate('click');
    expect(toggleEditing).toBeCalled();
  });

  it('clicking delete button should delete row', () => {
    const dataTable = new MockDataTable(twoRows);
    const data: MockData = {name: "Hello", number: 1};
    const toggleEditing = jest.fn();
    const buttons = mount(dataTable.getEditButtons(data, {...data}, false, toggleEditing));
    buttons.find('button[aria-label="Delete"]').simulate('click');

    expect(dataTable.removeData).toBeCalled();
    expect(dataTable.removeData).toBeCalledWith(data);
  });

  it('clicking cancel button on edited row should stop editing', () => {
    const dataTable = new MockDataTable(twoRows);
    const data: MockData = {name: "Hello", number: 1};
    const toggleEditing = jest.fn();
    const buttons = mount(dataTable.getEditButtons(data, {...data}, true, toggleEditing));
    buttons.find('button[aria-label="Cancel"]').simulate('click');

    expect(toggleEditing).toBeCalled();
  });

  it('should not save edited row when save button is clicked if incomplete', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.isDataComplete.mockReturnValue(false);

    const data: MockData = {name: "Hello", number: 1};
    const toggleEditing = jest.fn();
    const buttons = mount(dataTable.getEditButtons(data, {...data}, true, toggleEditing));
    buttons.find('button[aria-label="Save"]').simulate('click');
    expect(dataTable.isDataComplete).toBeCalled();
    expect(dataTable.isDataComplete).toBeCalledWith(data);
    expect(dataTable.isValid).not.toBeCalled();
    expect(dataTable.addData).not.toBeCalled();
  });

  it('clicking save button on edited row should not save if invalid', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.isValid.mockReturnValue(false);
    dataTable.isDataComplete.mockReturnValue(true);
    const data: MockData = {name: "Hello", number: 1};
    const toggleEditing = jest.fn();
    const buttons = mount(dataTable.getEditButtons(data, {...data}, true, toggleEditing));
    buttons.find('button[aria-label="Save"]').simulate('click');

    expect(dataTable.isDataComplete).toBeCalled();
    expect(dataTable.isDataComplete).toBeCalledWith(data);
    expect(dataTable.isValid).toBeCalled();
    expect(dataTable.isValid).toBeCalledWith(data);
    expect(dataTable.updateData).not.toBeCalled();
    expect(toggleEditing).not.toBeCalled();
  });

  it('clicking save button on edited row should save if valid', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.isValid.mockReturnValue(true);
    dataTable.isDataComplete.mockReturnValue(true);
    const oldValue: MockData = {name: "Hello", number: 1};
    const newValue: MockData = {name: "New Data", number: 2};
    const toggleEditing = jest.fn();
    const buttons = mount(dataTable.getEditButtons(oldValue, newValue, true, toggleEditing));
    buttons.find('button[aria-label="Save"]').simulate('click');

    expect(dataTable.isDataComplete).toBeCalled();
    expect(dataTable.isDataComplete).toBeCalledWith(newValue);
    expect(dataTable.isValid).toBeCalled();
    expect(dataTable.isValid).toBeCalledWith(newValue);
    expect(dataTable.updateData).toBeCalled();
    expect(dataTable.updateData).toBeCalledWith(newValue);
    expect(toggleEditing).toBeCalled();
  });

  it('should remove new row when cancel button is clicked', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.cancelAddingRow = jest.fn();

    const dataStore: any = {};
    const buttons = mount(dataTable.getAddButtons(dataStore));
    buttons.find('button[aria-label="Cancel"]').simulate('click');
    expect(dataTable.cancelAddingRow).toBeCalled();
  });

  it('should not save new row when save button is clicked if incomplete', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.isDataComplete.mockReturnValue(false);

    const dataStore: any = ['New Data', 1];
    const buttons = mount(dataTable.getAddButtons(dataStore));
    buttons.find('button[aria-label="Save"]').simulate('click');
    expect(dataTable.isDataComplete).toBeCalled();
    expect(dataTable.isDataComplete).toBeCalledWith(dataStore);
    expect(dataTable.isValid).not.toBeCalled();
    expect(dataTable.addData).not.toBeCalled();
  });

  it('should not save new row when save button is clicked if invalid', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.isDataComplete.mockReturnValue(true);
    dataTable.isValid.mockReturnValue(false);

    const dataStore: any = ['New Data', 1];
    const buttons = mount(dataTable.getAddButtons(dataStore));
    buttons.find('button[aria-label="Save"]').simulate('click');
    expect(dataTable.isDataComplete).toBeCalled();
    expect(dataTable.isDataComplete).toBeCalledWith(dataStore);
    expect(dataTable.isValid).toBeCalled();
    expect(dataTable.isValid).toBeCalledWith(dataStore);
    expect(dataTable.addData).not.toBeCalled();
  });

  it('should save new row when save button is clicked if valid', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.isDataComplete.mockReturnValue(true);
    dataTable.isValid.mockReturnValue(true);
    dataTable.cancelAddingRow = jest.fn();

    const dataStore: any = ['New Data', 1];
    const buttons = mount(dataTable.getAddButtons(dataStore));
    buttons.find('button[aria-label="Save"]').simulate('click');
    expect(dataTable.isDataComplete).toBeCalled();
    expect(dataTable.isDataComplete).toBeCalledWith(dataStore);
    expect(dataTable.isValid).toBeCalled();
    expect(dataTable.isValid).toBeCalledWith(dataStore);
    expect(dataTable.addData).toBeCalled();
    expect(dataTable.addData).toBeCalledWith(dataStore);
    expect(dataTable.cancelAddingRow).toBeCalled();
  });

  it('should pass getFilter to the FilterComponent', () => {
    const dataTable = shallow(<MockDataTable {...twoRows} />);
    const filter = jest.fn();
    (dataTable.instance() as MockDataTable).getFilter.mockReturnValue(filter);
    (dataTable.instance() as MockDataTable).forceUpdate();

    const filterComponent = dataTable.find(FilterComponent);
    const filterMap = filterComponent.props().filter;
    expect(filterMap).toEqual(filter);
  });

  it('should only render rows that match filter', () => {
    const dataTable = mount(<MockDataTable {...twoRows} />);
    (dataTable.instance() as MockDataTable).displayDataRow.mockClear();
    (dataTable.instance() as MockDataTable).editDataRow.mockClear();
    (dataTable.instance() as MockDataTable).setState({ currentFilter: v => v.number === 2 });

    expect((dataTable.instance() as MockDataTable).displayDataRow).toBeCalledTimes(1);
    expect((dataTable.instance() as MockDataTable).editDataRow).toBeCalledTimes(1);
    expect((dataTable.instance() as MockDataTable).displayDataRow)
      .toHaveBeenNthCalledWith(1, twoRows.tableData[1]);
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(1, twoRows.tableData[1], expect.any(Function));

    expect(toJson(dataTable)).toMatchSnapshot();
  });

  it('should have perPage set to 10 and page set to 1 initially', () => {
    const dataTable = new MockDataTable(twoRows);
    expect(dataTable.state.perPage).toEqual(10);
    expect(dataTable.state.page).toEqual(1);
  });

  it('should update page on set page', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.setState = jest.fn();
    dataTable.onSetPage({}, 3);
    expect(dataTable.setState).toBeCalled();
    expect(dataTable.setState).toBeCalledWith({ editedRows: [], page: 3 });
  });

  it('should update perPage on set per page', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.setState = jest.fn();
    dataTable.onPerPageSelect({}, 50);
    expect(dataTable.setState).toBeCalled();
    expect(dataTable.setState).toBeCalledWith(expect.any(Function));
    const value = (dataTable.setState as jest.Mock).mock.calls[0][0]({ page: 2, perPage: 25 });
    expect(value.editedRows).toEqual([]);
    expect(value.perPage).toEqual(50);
    expect(value.page).toEqual(1);
  });

  it('should only render data on page', () => {
    const dataTable = mount(<MockDataTable {...twoRows} />);
    (dataTable.instance() as MockDataTable).displayDataRow.mockClear();
    (dataTable.instance() as MockDataTable).editDataRow.mockClear();
    (dataTable.instance() as MockDataTable).setState({ perPage: 1, page: 1 });

    expect((dataTable.instance() as MockDataTable).displayDataRow).toBeCalledTimes(1);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[0]);
    expect((dataTable.instance() as MockDataTable).editDataRow).toBeCalledTimes(1);
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(1, twoRows.tableData[0], expect.any(Function));

    (dataTable.instance() as MockDataTable).displayDataRow.mockClear();
    (dataTable.instance() as MockDataTable).editDataRow.mockClear();
    (dataTable.instance() as MockDataTable).setState({ perPage: 1, page: 2 });

    expect((dataTable.instance() as MockDataTable).displayDataRow).toBeCalledTimes(1);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[1]);
    expect((dataTable.instance() as MockDataTable).editDataRow).toBeCalledTimes(1);
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(1, twoRows.tableData[1], expect.any(Function));

    expect(toJson(dataTable)).toMatchSnapshot();
  });

  it('should update sortBy on sort', () => {
    const dataTable = new MockDataTable(twoRows);
    dataTable.setState = jest.fn();
    dataTable.onSort({}, 0, "asc");
    expect(dataTable.setState).toBeCalled();
    expect(dataTable.setState).toBeCalledWith({ editedRows: [], sortBy: { index: 0, direction: "asc" } });
  });

  it('should render rows in sorted order when ascending', () => {
    const dataTable = mount(<MockDataTable {...twoRows} tableData={[twoRows.tableData[1], twoRows.tableData[0]]} />);
    (dataTable.instance() as MockDataTable).displayDataRow.mockClear();
    (dataTable.instance() as MockDataTable).editDataRow.mockClear();
    (dataTable.instance() as MockDataTable).onSort({}, 1, "asc");

    expect((dataTable.instance() as MockDataTable).displayDataRow).toBeCalledTimes(2);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[0]);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(2, twoRows.tableData[1]);
    expect((dataTable.instance() as MockDataTable).editDataRow).toBeCalledTimes(2);
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(1, twoRows.tableData[0], expect.any(Function));
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(2, twoRows.tableData[1], expect.any(Function));

    expect(toJson(dataTable)).toMatchSnapshot();
  });

  it('should render rows in reverse of sorted order when descending', () => {
    const dataTable = mount(<MockDataTable {...twoRows} />);
    (dataTable.instance() as MockDataTable).displayDataRow.mockClear();
    (dataTable.instance() as MockDataTable).editDataRow.mockClear();
    (dataTable.instance() as MockDataTable).onSort({}, 1, "desc");

    expect((dataTable.instance() as MockDataTable).displayDataRow).toBeCalledTimes(2);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(1, twoRows.tableData[1]);
    expect((dataTable.instance() as MockDataTable).displayDataRow).toHaveBeenNthCalledWith(2, twoRows.tableData[0]);
    expect((dataTable.instance() as MockDataTable).editDataRow).toBeCalledTimes(2);
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(1, twoRows.tableData[1], expect.any(Function));
    expect((dataTable.instance() as MockDataTable).editDataRow)
      .toHaveBeenNthCalledWith(2, twoRows.tableData[0], expect.any(Function));

    expect(toJson(dataTable)).toMatchSnapshot();
  });
});

const noRows: DataTableProps<MockData> = {
  ...useTranslation(),
  tReady: true,
  title: "Data Table",
  columnTitles: ["Column 1", "Column 2"],
  tableData: [],
};

const twoRows: DataTableProps<MockData> = {
  ...useTranslation(),
  tReady: true,
  title: "Data Table",
  columnTitles: ["Column 1", "Column 2"],
  tableData: [{name: "Some Data", number: 1}, {name: "More Data", number: 2}],
};
