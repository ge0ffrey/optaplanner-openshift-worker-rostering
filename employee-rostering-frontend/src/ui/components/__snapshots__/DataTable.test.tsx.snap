// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable component should render correctly with a few rows 1`] = `
<div>
  <Button
    onClick={[Function]}
  >
    Add
  </Button>
  <Table
    borders={true}
    caption="Data Table"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Column 1",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Column 2",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onCollapse={null}
    onExpand={null}
    rowLabeledBy="simple-node"
    rows={
      Array [
        Object {
          "cells": Array [
            "Some Data",
            1,
          ],
          "props": Object {},
        },
        Object {
          "cells": Array [
            "More Data",
            2,
          ],
          "props": Object {},
        },
      ]
    }
    variant={null}
  >
    <TableHeader
      className=""
    />
    <TableBody
      className=""
      onRowClick={[Function]}
      rowKey="id"
    />
  </Table>
</div>
`;

exports[`DataTable component should render correctly with no rows 1`] = `
<div>
  <Button
    onClick={[Function]}
  >
    Add
  </Button>
  <Table
    borders={true}
    caption="Data Table"
    cells={
      Array [
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Column 1",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "Column 2",
        },
        Object {
          "cellTransforms": Array [
            [Function],
          ],
          "props": Object {},
          "title": "",
        },
      ]
    }
    className=""
    contentId="expanded-content"
    dropdownDirection="down"
    dropdownPosition="right"
    expandId="expandable-toggle"
    gridBreakPoint="grid-md"
    onCollapse={null}
    onExpand={null}
    rowLabeledBy="simple-node"
    rows={Array []}
    variant={null}
  >
    <TableHeader
      className=""
    />
    <TableBody
      className=""
      onRowClick={[Function]}
      rowKey="id"
    />
  </Table>
</div>
`;
